agora/Agora/Governor/Scripts.hs

```diff
@@ -489,35 +496,44 @@ governorValidator as =
           pguardC "Output GATs is more than minted GATs" $
             plength # outputsWithGAT #== gatCount
 
-          let gatOutputValidator' :: Term s (PMap _ PValidatorHash PDatumHash :--> PTxOut :--> PBool)
-              gatOutputValidator' =
+          let validateGATOutput' :: Term s (PProposalEffectGroup :--> PTxOut :--> PBool)
+              validateGATOutput' =
                 phoistAcyclic $
                   plam
-                    ( \effects output' -> unTermCont $ do
-                        output <- pletFieldsC @'["address", "datum"] output'
-
-                        let scriptHash =
-                              passertPJust # "GAT receiver is not a script"
-                                #$ pscriptHashFromAddress # output.address
-                            datumHash =
-                              ptrace
-                                "Output to effect should have datum"
-                                pfromDatumHash
-                                # output.datum
-
-                            expectedDatumHash =
-                              passertPJust # "Receiver is not in the effect list"
-                                #$ plookup # scriptHash # effects
+                    ( \effects output -> unTermCont $ do
+                        outputF <- pletFieldsC @'["address", "datum", "referenceScript"] output
+
+                        let receiverScriptHash =
+                              passertPJust # "GAT receiver should be a script"
+                                #$ pscriptHashFromAddress # outputF.address
+                            effect =
+                              passertPJust # "Receiver should be in the effect group"
+                                #$ AssocMap.plookup # receiverScriptHash # effects
+                            hasCorrectReferenceScript =
+                              pmaybeData
+                                # pconstant True
+                                # plam
+                                  ( ( passertPDJust
+                                        # "Output UTXO should have a reference script"
+                                        # outputF.referenceScript
+                                        #==
+                                    )
+                                      . pfromData
+                                  )
+                                # (psndTuple # effect)
```

Comment:
Looks great! I am still slightly skeptical of the design (not the implementation), so we may want to change this in the future to _not_ use reference scripts, and use just a script has instead. I'm unsure how to encode that in a backwards compatible way though. This approach right now is backwards compatible with not using reference scripts at all, for example. Maybe we'll need to encode the hash in the `TokenName` instead? Some things to think about.

The most troubling thing is really the script size.

Reply:
We should really turn off all the tracing stuff for benchmarking purposes, but unfortunately our current setup can't do this.